README.md
# DiepTD
A repository for Diep.io Tower Defense, a tower defense (or real time strategy) game based on Diep.io, a 2D shooter.


changelog.txt
Update 0.1.8.1: Fixed a bug where completing levels doesn't make them show as complete.
Update 0.1.8: Levels 2 and 3 are now released!
A crapton of bug fixes.
Some other stuff that I forgot.
Update 0.1.7: TUTORIAL LEVEL COMPLETE!
Lots of bug fixes.
Level selection works now.
Stay tuned for more levels!
Update 0.1.6.1: Fixed pentashot bug. Now, pressing the [1] key upgrades a triple shot to a pentashot, not a triple shot to a twin.
Similar bug fixed with twin upgrading to triple shot with the [1] shortcut. It used to not upgrade at all, but now it upgrades to triple shot.
Update 0.1.6: LEVELS HAVE BEEN ADDED! So far, the level system is in place and will probably work semi-effectively.
Tanks can now be recycled.
Negative money no longer works.
Tanks/Enemies cannot be spawned with random keys now.
Text info changed significantly.
Note that many of these changes are temporary. I'll be adding a debug option soon if you want to mess around with the game.
Update 0.1.5.2: Some changes to power system, more graphical changes, bug fixes, that kind of stuff.
Working on documentation and showing what stuff does and all that.
Update 0.1.5.1: Changed the graphics of tank FOV. You can change the FOV to the original if you want.
Update 0.1.5: HUGE UPDATE! New Extractor tank, it extracts power from polygons to generate electricty and power the other tanks.
Extractor tank still work in progress, selection stuff isn't done.
Tank connectors completed.
Power system completed (Comment from developer (adasba): Huge pain in the ass, took me HOURS!)
Graphical changes
When placing a tank, all connectors turn green so you know where you can place it.
Bug fixes.
Tank placement is now restricted to areas close to connectors.
Update 0.1.4.2: Tank connectors, they connect stuff.
Update 0.1.4.1: Instructions changed, "E" now spawns a penta shot.
Update 0.1.4: Pentashot tank added, by far the most powerful one.
New icons for spawning tanks and miners.
Some other stuff I forgot to write down earlier.
Update 0.1.3.2: Accidentally changed enemy tank texture. This has been fixed.
Update 0.1.3.1: Graphical changes to polygon miner.
Update 0.1.3: Currency added to the game, called "Resources".
Some changes to upgrade icons and tank spawning.
New polygon miner tank to mine resources from shapes.
Shapes have health and can be damaged.
Tanks regenerate.
Update 0.1.2.1: Upgrade icons can now be clicked on, and their animations were slightly modified.
Tank selection settings changed, now they don't de-select when you click somewhere else.
Press X or Escape to exit from the tank upgrade screen.
Update 0.1.2: Tri-Angle added, upgrade icons are complete for now.
Update 0.1.1.2: Much more code condensing, the game is down to 828 lines, it was at around 1200 before.
Update 0.1.1.1: Working on upgrade graphics, several graphical changes, and condensing code.
Update 0.1.1: Added the ability to upgrade tanks by clicking on them and pressing a number key.
Added blue machine gun.
Update 0.1: Started recording updates in a changelog.
Update idk: why are you reading this, go away

scripts/init.js
//context and canvas
var c = document.getElementById("myCanvas");
var ctx = c.getContext("2d");
//x and y position
var x = 0;
var y = 0;
//tank variables
var xS = [];
var yS = [];
var tS = [];
var hPS = [];
var mHPS = [];
var tMRS = [];
var tMRS2 = [];
var dS = [];
var pW = [];
var pW2 = [];
var costS = [];
//some mouse crap
var pMX = 0;
var pMY = 0;
var mP = {};
var mID = 0;
//"counter" variables that count stuff
var lctr = 0;
var ctr = 0;
var ctr2 = 0;
var ctr3 = 0;
var ctr4 = 0;
var ctr5 = 0;
var ctr6 = 0;
var ctr7 = 0;
var ctr8 = 0;
var ctr9 = 0;
var ctr10 = 0;
var ctr11 = 0;
var ctr12 = 0;
var ctr13 = 0;
var ctrS = [0, 0, 0, 0, 0];
//idk what these are, they're basically just for random tanks. aN is anti-nuisance.
var aN = 0;
var aN2 = [];
//polygon variables
var xP = [];
var yP = [];
var tP = [];
var hPP = [];
var mHPP = [];
//enemy variables
var xE = [];
var yE = [];
var tE = [];
var dE = [];
var hPE = [];
var mHPE = [];
var tMRE = [];
var tMRE2 = [];
//bullet variables
var xB = [];
var yB = [];
var dB = [];
var lTB = [];
var teams = [];
var vB = [];
var tB = [];
//used for determining what a tank/enemy will aim at
var cEV = 0;
var cEV2 = 0;
var cEV3 = 0;
var cED = 0;
var cED2 = 0;
//used to determining whether a key was pressed
var keys = Array(256);
var keys2 = Array(256);
//used for selecting and adding new towers
var sT = 0;
var sT2 = 0;
var sTV = 0;
//cash and power variables, half of these I'm not using or something
var cash = 100;
var power = 0;
var mPower = 0;
//the type of tank that is selected to be placed
var pT = 0;
//prevents multiple clicks from auto-upgrading a tank a bunch of times
var aRCV = 0;
//output is dependant on whether your mouse is within an area where a tank can be placed
var iR = 0;
//more power stuff
var pWC = 0;
var pWL = [];
var totalPower = 0;
//fov style, changes a graphical setting
var FOVStyle = 1;
//stands for level active, and it is on when a level is being played. "level" is self-explanatory. levelC is "level complete"
var lVA = 0;
var level = 0;
var levelC = [0, 0, 0];

scripts/render.js
//barrel rendering script
function renderBarrel(x, y, l, w, offset, dir) {
    ctx.beginPath();
    ctx.fillStyle = "#555555";
    ctx.moveTo((offset + w / 2) * Math.cos(dir - Math.PI / 2) + x, (offset + w / 2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + w / 2) * Math.cos(dir - Math.PI / 2) - l * Math.sin(dir - Math.PI / 2) + x, (offset + w / 2) * Math.sin(dir - Math.PI / 2) + l * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + w / -2) * Math.cos(dir - Math.PI / 2) - l * Math.sin(dir - Math.PI / 2) + x, (offset + w / -2) * Math.sin(dir - Math.PI / 2) + l * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + w / -2) * Math.cos(dir - Math.PI / 2) + x, (offset + w / -2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = "#999999";
    ctx.moveTo(((offset - 3) + w / 2) * Math.cos(dir - Math.PI / 2) + x, ((offset - 3) + w / 2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset - 3) + w / 2) * Math.cos(dir - Math.PI / 2) - (l - 3) * Math.sin(dir - Math.PI / 2) + x, ((offset - 3) + w / 2) * Math.sin(dir - Math.PI / 2) + (l - 3) * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset + 3) + w / -2) * Math.cos(dir - Math.PI / 2) - (l - 3) * Math.sin(dir - Math.PI / 2) + x, ((offset + 3) + w / -2) * Math.sin(dir - Math.PI / 2) + (l - 3) * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset + 3) + w / -2) * Math.cos(dir - Math.PI / 2) + x, ((offset + 3) + w / -2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.fill();
    ctx.closePath();
}
//shadow barrel rendering script
function renderShadowBarrel(x, y, l, w, offset, dir) {
    ctx.beginPath();
    ctx.fillStyle = "rgb(110, 110, 110)";
    ctx.moveTo((offset + (w / 2) * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / 2) * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (w / 2) * 1.15) * Math.cos(dir - Math.PI / 2) - (l * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / 2) * 1.1) * Math.sin(dir - Math.PI / 2) + (l * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (w / -2) * 1.15) * Math.cos(dir - Math.PI / 2) - (l * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / -2) * 1.1) * Math.sin(dir - Math.PI / 2) + (l * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (w / -2) * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / -2) * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.fill();
    ctx.closePath();
}
//trapezoid barrel, self explanatory
function renderTrapezoidBarrel(x, y, l, w, offset, dir, outerWidth) {
    ctx.beginPath();
    ctx.fillStyle = "#555555";
    ctx.moveTo((offset + w / 2) * Math.cos(dir - Math.PI / 2) + x, (offset + w / 2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + outerWidth / 2) * Math.cos(dir - Math.PI / 2) - l * Math.sin(dir - Math.PI / 2) + x, (offset + outerWidth / 2) * Math.sin(dir - Math.PI / 2) + l * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + outerWidth / -2) * Math.cos(dir - Math.PI / 2) - l * Math.sin(dir - Math.PI / 2) + x, (offset + outerWidth / -2) * Math.sin(dir - Math.PI / 2) + l * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo((offset + w / -2) * Math.cos(dir - Math.PI / 2) + x, (offset + w / -2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = "#999999";
    ctx.moveTo(((offset - 3) + w / 2) * Math.cos(dir - Math.PI / 2) + x, ((offset - 3) + w / 2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset - 3) + outerWidth / 2) * Math.cos(dir - Math.PI / 2) - (l - 3) * Math.sin(dir - Math.PI / 2) + x, ((offset - 3) + outerWidth / 2) * Math.sin(dir - Math.PI / 2) + (l - 3) * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset + 3) + outerWidth / -2) * Math.cos(dir - Math.PI / 2) - (l - 3) * Math.sin(dir - Math.PI / 2) + x, ((offset + 3) + outerWidth / -2) * Math.sin(dir - Math.PI / 2) + (l - 3) * Math.cos(dir - Math.PI / 2) + y);
    ctx.lineTo(((offset + 3) + w / -2) * Math.cos(dir - Math.PI / 2) + x, ((offset + 3) + w / -2) * Math.sin(dir - Math.PI / 2) + y);
    ctx.fill();
    ctx.closePath();
}
//i don't need to provide another description, this one is obvious
function renderTrapezoidShadowBarrel(x, y, l, w, offset, dir, outerWidth) {
    ctx.beginPath();
    ctx.fillStyle = "rgb(110, 110, 110)";
    ctx.moveTo((offset + (w / 2) * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / 2) * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (outerWidth / 2) * 1.15) * Math.cos(dir - Math.PI / 2) - (l * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (outerWidth / 2) * 1.1) * Math.sin(dir - Math.PI / 2) + (l * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (outerWidth / -2) * 1.15) * Math.cos(dir - Math.PI / 2) - (l * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (outerWidth / -2) * 1.1) * Math.sin(dir - Math.PI / 2) + (l * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.lineTo((offset + (w / -2) * 1.15) * Math.cos(dir - Math.PI / 2) + 512 + (x - 512) * 1.1, (offset + (w / -2) * 1.15) * Math.sin(dir - Math.PI / 2) + 512 + (y - 512) * 1.1);
    ctx.fill();
    ctx.closePath();
}
//adds a bullet
function summonBullet(x, y, dir, lifetime, team, vel, tank, offset) {
    if (offset == undefined) {
        xB.push(x);
        yB.push(y);
    } else {
        xB.push(x + Math.cos(dir + Math.PI / 2) * offset);
        yB.push(y + Math.sin(dir + Math.PI / 2) * offset);
    }
    dB.push(dir);
    lTB.push(lifetime);
    teams.push(team);
    vB.push(vel);
    tB.push(tank);
}
//adds a tower
function summonTower(x, y, type) {
    xS.push(x);
    yS.push(y);
    tS.push(type);
    tMRS.push(0);
    tMRS2.push(0);
    dS.push(0);
    if (type == 8) {
        pW.push(0);
        pW2.push(new Array());
    } else {
        pW.push(new Array());
        pW2.push(0);
    }
    if (type == 0) {
        hPS.push(150);
        mHPS.push(150);
        costS.push(30);
    }
    if (type == 1) {
        hPS.push(225);
        mHPS.push(225);
    }
    if (type == 2) {
        hPS.push(275);
        mHPS.push(275);
    }
    if (type == 3) {
        hPS.push(250);
        mHPS.push(250);
    }
    if (type == 4) {
        hPS.push(80);
        mHPS.push(80);
        costS.push(20);
    }
    if (type == 5) {
        hPS.push(350);
        mHPS.push(350);
    }
    if (type == 6) {
        hPS.push(500);
        mHPS.push(500);
    }
    if (type == 7) {
        hPS.push(50);
        mHPS.push(50);
        costS.push(10);
    }
    if (type == 8) {
        hPS.push(750);
        mHPS.push(750);
        costS.push(50);
    }
}
//adds an enemy
function summonEnemy(x, y, type) {
    xE.push(x);
    yE.push(y);
    tE.push(type);
    tMRE.push(0);
    tMRE2.push(0);
    dE.push(0);
    if (type == 0) {
        hPE.push(30);
        mHPE.push(30);
    } else if (type == 1) {
        hPE.push(40);
        mHPE.push(40);
    } else if (type == 2) {
        hPE.push(50);
        mHPE.push(50);
    } else if (type == 3) {
        hPE.push(35);
        mHPE.push(35);
    } else if (type == 4) {
        hPE.push(55);
        mHPE.push(55);
    }
}
function summonPolygon(x, y, type) {
    xP.push(x);
    yP.push(y);
    tP.push(type);
    if (type == 0) {
        hPP.push(1000);
        mHPP.push(1000);
    } else if (type == 1) {
        hPP.push(2000);
        mHPP.push(2000);
    } else if (type == 2) {
        hPP.push(10000);
        mHPP.push(10000);
    }
}
//renders a polygon
function polygon(x, y, sides, size, dir, color) {
    ctr6 = 0;
    ctx.beginPath();
    ctx.fillStyle = color;
    ctx.moveTo(x + Math.cos(dir) * size, y + Math.sin(dir) * size);
    while (ctr6 < sides) {
        ctx.lineTo(x + Math.cos(dir + ctr6 * (2 * Math.PI) / sides) * size, y + Math.sin(dir + ctr6 * (2 * Math.PI) / sides) * size);
        ctr6++;
    }
    ctx.fill();
    ctx.closePath();
}
//renders a health bar
function healthBar(x, y, hp, maxHp, radius) {
    ctx.beginPath();
    ctx.strokeStyle = "rgba(85, 85, 85, " + ((0.35 / maxHp) * (maxHp - hp) + 0.05) + ")";
    ctx.lineWidth = 11
    ctx.arc(x, y, radius, 0, Math.PI * 2)
    ctx.stroke();
    ctx.closePath();
    ctx.beginPath();
    ctx.lineWidth = 5
    ctx.strokeStyle = "rgba(106, " + Math.round((182 / maxHp) * hp) + ", " + Math.round((100 / maxHp) * hp) + ", " + ((0.875 / maxHp) * (maxHp - hp) + 0.125) + ")";
    ctx.arc(x, y, radius, 0, ((Math.PI * 2) / maxHp) * hp);
    ctx.stroke();
    ctx.closePath();
}
//renders a "diep circle" with a border that has the color "#555555"
function diepCircle(x, y, radius, color) {
    ctx.beginPath();
    ctx.fillStyle = "#555555";
    ctx.arc(x, y, radius, 0, Math.PI * 2);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = color;
    ctx.arc(x, y, radius - 3, 0, Math.PI * 2);
    ctx.fill();
    ctx.closePath();
}
//diepcircle without the border. Literally just a circle.
function diepCircle2(x, y, radius, color) {
    ctx.beginPath();
    ctx.fillStyle = color;
    ctx.arc(x, y, radius, 0, Math.PI * 2);
    ctx.fill();
    ctx.closePath();
}
//field of view, the thing that shows how far a tank can shoot/reach
function FOV(x, y, radius, color, color2) {
    ctx.beginPath();
    ctx.fillStyle = color;
    ctx.arc(x, y, radius, 0, 2 * Math.PI);
    ctx.fill();
    ctx.closePath();
    if (FOVStyle == 0) {
        ctx.beginPath();
        ctx.lineWidth = 5;
        ctx.strokeStyle = color2;
        ctx.arc(x, y, radius - 2.5, 0, 2 * Math.PI);
        ctx.stroke();
        ctx.closePath();
    }

}
//a tank icon, which contains info like a tank upgrade
function tankIcon(x, y, color1, color2) {
    ctx.beginPath();
    ctx.fillStyle = color1;
    ctx.lineWidth = 7;
    ctx.fillRect(x, y, 150, 85);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = color2;
    ctx.lineWidth = 7;
    ctx.fillRect(x, y + 85, 150, 65);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.strokeStyle = "#555555";
    ctx.lineWidth = 5;
    ctx.strokeRect(x, y, 150, 150);
    ctx.stroke();
    ctx.closePath();
}
function tankIcon3(x, y, color1, color2) {
    ctx.beginPath();
    ctx.fillStyle = color1;
    ctx.lineWidth = 7;
    ctx.fillRect(x - 50, y, 200, (100 / 30) * 17);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = color2;
    ctx.lineWidth = 7;
    ctx.fillRect(x - 50, y + (100 / 30) * 17, 200, (100 / 30) * 13);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.strokeStyle = "#555555";
    ctx.lineWidth = 5;
    ctx.strokeRect(x - 50, y, 200, 100);
    ctx.stroke();
    ctx.closePath();
}
//smaller tank icon that is usually used to display text
function tankIcon2(x, y, color1, color2) {
    ctx.beginPath();
    ctx.fillStyle = color1;
    ctx.lineWidth = 7;
    ctx.fillRect(x, y, 150, 17);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.fillStyle = color2;
    ctx.lineWidth = 7;
    ctx.fillRect(x, y + 17, 150, 13);
    ctx.fill();
    ctx.closePath();
    ctx.beginPath();
    ctx.strokeStyle = "#555555";
    ctx.lineWidth = 5;
    ctx.strokeRect(x, y, 150, 30);
    ctx.stroke();
    ctx.closePath();
}

scripts/utility.js
//does calculations for "powering" a tank
function powerTank(tank, power1) {
    ctr7 = 0;
    aN = 0;
    aN2 = [];
    while (pW[tank].length > ctr7) {
        if (pW[pW[tank][ctr7]] >= power1 / (pW[tank].length - aN)) {
            aN2.push(pW[tank][ctr7]);
        } else {
            aN++;
        }
        ctr7++;
    }
    ctr7 = 0;
    while (aN2.length > ctr7) {
        pW[aN2[ctr7]] = pW[aN2[ctr7]] - (power1 / (pW[tank].length - aN));
        ctr7++;
    }
}
//sees how much usable power there is for a tank
function usableTankPower(tank) {
    aN = 0;
    ctr7 = 0;
    while (pW[tank].length > ctr7) {
        aN = aN + pW[pW[tank][ctr7]]
        ctr7++;
    }
    return aN;
}
//makes recycle buttons with less lines of code.
function recycle(variable1) {
    ctx.textAlign = "center";
    tankIcon2(50, 784 + variable1, "#CA7676", "#A56262");
    ctx.font = "Bold 14px Ubuntu";
    ctx.fillStyle = "White";
    ctx.lineWidth = 4;
    ctx.strokeStyle = "rgb(40, 40, 40)";
    ctx.strokeText("Recycle [R]", 125, 805 + variable1);
    ctx.fillText("Recycle [R]", 125, 805 + variable1);
}
//ends a level once there are no tanks
function endLevel() {
    lVA = 0;
    lctr = 0;
    xE = [];
    yE = [];
    tE = [];
    dE = [];
    hPE = [];
    mHPE = [];
    tMRE = [];
    tMRE2 = [];
    xB = [];
    yB = [];
    dB = [];
    lTB = [];
    teams = [];
    vB = [];
    tB = [];
    xP = [];
    yP = [];
    tP = [];
    hPP = [];
    mHPP = [];
    xE = [];
    yE = [];
    tE = [];
    dE = [];
    hPE = [];
    mHPE = [];
    tMRE = [];
    tMRE2 = [];
    level = -1;
    ctr9 = 0;
    ctr12 = 0;
    xS = [];
    yS = [];
    tS = [];
    var hPS = [];
    var mHPS = [];
    var tMRS = [];
    var tMRS2 = [];
    var dS = [];
    var pW = [];
    var pW2 = [];
    var costS = [];
}
function deathMsg() {
    ctr8 = 1;
    deathM = deathMS[Math.floor(Math.random() * deathMS.length)];
}
function message(message, message2) {
    ctr11 = 1;
    msg = message;
    if (message2 != undefined || message2 != "") {
        msg2 = message2;
    }
}
function startLevel(whichLevel) {
    ctr10 = 0;
    while (ctr10 < sConfig[whichLevel].length) {
        if (sConfig[whichLevel][ctr10].type == "square") {
            summonPolygon(sConfig[whichLevel][ctr10].x, sConfig[whichLevel][ctr10].y, 0);
        } else if (sConfig[whichLevel][ctr10].type == "triangle") {
            summonPolygon(sConfig[whichLevel][ctr10].x, sConfig[whichLevel][ctr10].y, 1);
        } else if (sConfig[whichLevel][ctr10].type == "pentagon") {
            summonPolygon(sConfig[whichLevel][ctr10].x, sConfig[whichLevel][ctr10].y, 2);
        } else if (sConfig[whichLevel][ctr10].type == "tank") {
            summonTower(sConfig[whichLevel][ctr10].x, sConfig[whichLevel][ctr10].y, sConfig[whichLevel][ctr10].tankType);
        }
        ctr10++
    }
    ctr10 = 0;
}
function levelIcon(whatLevel, description, difficulty, iconX, iconY) {
    tankIcon3(iconX - x + 512, iconY - y + 512, "#75C6C3", "#60A2A0");
    ctx.fillStyle = "White";
    ctx.lineWidth = 6;
    ctx.strokeStyle = "rgb(40, 40, 40)";
    ctx.font = "Bold 14px Ubuntu";
    ctx.textAlign = "center";
    ctx.strokeText("Level " + (whatLevel + 1), iconX + 50 - x + 512, iconY + 20 - y + 512);
    ctx.fillText("Level " + (whatLevel + 1), iconX + 50 - x + 512, iconY + 20 - y + 512);
    ctx.strokeText(description, iconX + 50 - x + 512, iconY + 43 - y + 512);
    ctx.fillText(description, iconX + 50 - x + 512, iconY + 43 - y + 512);
    ctx.strokeText(difficulty, iconX + 50 - x + 512, iconY + 66 - y + 512);
    ctx.fillText(difficulty, iconX + 50 - x + 512, iconY + 66 - y + 512);
    if (levelC[whatLevel] == 0) {
        ctx.strokeText("Incomplete", iconX + 50 - x + 512, iconY + 89 - y + 512);
        ctx.fillText("Incomplete", iconX + 50 - x + 512, iconY + 89 - y + 512);
    } else {
        ctx.strokeText("Completed!", iconX + 50 - x + 512, iconY + 89 - y + 512);
        ctx.fillText("Completed!", iconX + 50 - x + 512, iconY + 89 - y + 512);
    }
}
function lvlMouseBox(boxX, boxY) {
    if (mP.x > -50 - x + 512 + boxX && mP.y > 0 - y + 512 + boxY && mP.x < 150 - x + 512 + boxX && mP.y < 100 - y + 512 + boxY) {
        return true;
    } else {
        return false;
    }
}

scripts/config.js
//death message generation
var deathM;
var deathMS = ["Rest in spaghetti, never forghetti...", "Ripperoni.", "Killed to death!", "Git gud m8!", "OMG HACKERS!1!1!!", "Wubba Lubba Dub Dub!", "Adasba, fix your game...", "[Insert witty death message here]"];
//ordinary message stuff
var msg;
var msg2;
var msgS = [[{ message: "Welcome to Diep TD!", time: 0, duration: 300 },
    { message: "Diep TD is a real time strategy game where you place tanks to fend off enemies.", time: 300, duration: 500 },
    { message: "Find the button for the miner, and click it.", time: 800, duration: 600 },
    { message: "Click again to place the miner. You can only place it in the green area.", message2: "Make sure the pentagon is within the miner's grey circle.", time: 1400, duration: 1200 },
    { message: "Look at the resources bar in the top right corner.", message2: "The miner is converting the pentagon into resources you can use to build tanks.", time: 2600, duration: 1000 },
    { message: "A single miner may not be enough. Place down a few more. The [M] key is an alternative to the button.", message2:"If you run out of space to place miners, use a connector, hotkey is [C].", time: 3600, duration: 600 },
    { message: "Enemy tanks will be approaching soon.", message2: "Find the button for the tank, and place it to the right of the big green object, which is called an extractor.", time: 5000, duration: 1200 },
    { message: "Even more enemies are attacking. A single tank will not do the job. The [SPACE] key may speed up tank placement.", message2: "You can place more tanks, or click on the already-placed tank to see the possible upgrades.", time: 7200, duration: 1200 },
    { message: "Enemy tanks will soon attack from all directions. You will need to place tanks all around your miners and extractor.", message2: "If you cannot place tanks where you want, use a connector (hotkey is [C]) to add extra tank placement area.", time: 8400, duration: 1200 },
    { message: "Just one extractor may not be enough to handle the power drain from your tanks.", message2: "Use the extractor button or the [E] key to add one near the red triangle.", time: 9600, duration: 1200 },
    { message: "30 enemy tanks are now approaching from all directions. Don't panic if you lose some tanks, they can be replaced.", message2: "If you run out of money while under attack, you can use the [R] key or the recycle button to sell something you don't need.", time: 10800, duration: 1400 },
    { message: "There are two waves of enemies left. You're on your own now.", message2: "Good luck!", time: 12500, duration: 600 },
    { message: "It looks like I was wrong. There is a third wave coming. This one is far too powerful to fight off.", message2: "By holding off the enemy, you've given us enough time to send you rescue tanks. Until then, try to defend!", time: 18000, duration: 1200 },
    { message: "In a few seconds we will be there.", message2: "Survive until then.", time: 19000, duration: 500 },
    { message: "Level 1 Complete!", message2: "Levels 2 and 3 unlocked!", time: 19500, duration: 500 }],
    [{ message: "The enemy seems to be aware of our position now.", time: 0, duration: 400 },
    { message: "They'll be sending everything they have at us, including that huge army of tri-angles we faced earlier.", time: 400, duration: 800 },
    { message: "Enemies approaching from the right!", time: 4500, duration: 300 },
    { message: "Enemies approaching from the bottom!", message2: "Looks like they're approaching in a clockwise direction. Remember, sell tanks you aren't using!", time: 6000, duration: 1000 },
    { message: "The huge army of tri-angles are approaching again.", message2: "You should be able to fight them off... hopefully...", time: 19000, duration: 1000 },
    { message: "You know what? Let me help you with this. Get as many tanks as possible, and some extractors.", message2: "When you run out of money, sell your miners. This is the last wave, so selling them will give you more money than if you use them.", time: 20000, duration: 1200 }],
    [{ message: "One of our bases has sustained damage, and it needs to be rebuilt.", time: 0, duration: 800 },
    { message: "Due to a lack of resources, we've sent you to mine some squares.", message2: "This is enemy territory, so stay alert. The more time you take, the stronger they get.", time: 800, duration: 1200 }
    ]];
var msgD;
//variables for level data. "sConfig" is "starting configuration". "eData" is "enemyData"
var sConfig = [[{ type: "triangle", x: 150, y: 0 }, { type: "pentagon", x: -300, y: -100 }, { type: "tank", x: 0, y: 0, tankType: 8 }],
    [{ type: "square", x: 322, y: -176 }, { type: "triangle", x: -437, y: 100 }, { type: "square", x: 50, y: 50 }, { type: "square", x: 180, y: 515 }, { type: "tank", x: 0, y: 0, tankType: 8 }],
    [{ type: "square", x: 700, y: 700 }, { type: "square", x: -700, y: 700 }, { type: "square", x: 700, y: -700 }, { type: "square", x: -700, y: -700 }, { type: "triangle", x: 50, y: 0 }, { type: "tank", x: 0, y: 0, tankType: 8 }]];
var eData = [[{ type: 0, x: 2500, y: 0, time: 5000, pattern: 0, },
    { type: 0, x: 2500, y: -1250, time: 6400, count: 10, pattern: 1, ySpacing: 250, xSpacing: 0 },
    { type: 0, x: 0, y: 0, time: 10800, count: 30, pattern: 2, dirSpacing: Math.PI / 15, radius: 2500, startDir: 0 },
    { type: 2, x: 0, y: 0, time: 13000, count: 30, pattern: 2, dirSpacing: Math.PI / 15, radius: 2500, startDir: 0 },
    { type: 3, x: -2500, y: -500, time: 15500, count: 20, pattern: 1, ySpacing: 50, xSpacing: 0 },
    { type: 4, x: 0, y: 0, time: 18000, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 1500, startDir: 0 },
    { type: 4, x: 0, y: 0, time: 18300, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 1500, startDir: Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 18600, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 1500, startDir: 2 * Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 18900, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 1500, startDir: 3 * Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 19200, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 1500, startDir: 4 * Math.PI / 100 }],
    [{ type: 1, x: 2000, y: -750, time: 4500, count: 4, pattern: 1, ySpacing: 500, xSpacing: 0 },
    { type: 0, x: -625, y: 2000, time: 6000, count: 10, pattern: 1, ySpacing: 0, xSpacing: 250 },
    { type: 0, x: -2000, y: -850, time: 7500, count: 16, pattern: 1, ySpacing: 100, xSpacing: 0 },
    { type: 2, x: -1100, y: -2000, time: 9000, count: 10, pattern: 1, ySpacing: 0, xSpacing: 200 },
    { type: 3, x: 2000, y: -650, time: 10500, count: 12, pattern: 1, ySpacing: 100, xSpacing: 0 },
    { type: 0, x: -575, y: 2000, time: 12000, count: 24, pattern: 1, ySpacing: 0, xSpacing: 50 },
    { type: 4, x: -2000, y: -1700, time: 13500, count: 16, pattern: 1, ySpacing: 200, xSpacing: 0 },
    { type: 1, x: -775, y: -2000, time: 15000, count: 30, pattern: 1, ySpacing: 0, xSpacing: 50 },
    { type: 4, x: 0, y: 0, time: 20000, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 3000, startDir: 0 },
    { type: 4, x: 0, y: 0, time: 20750, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 3000, startDir: Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 21500, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 3000, startDir: 2 * Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 22250, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 3000, startDir: 3 * Math.PI / 100 },
    { type: 4, x: 0, y: 0, time: 23000, count: 40, pattern: 2, dirSpacing: Math.PI / 20, radius: 3000, startDir: 4 * Math.PI / 100 }],
    [{ type: 0, x: 4000, y: 3000, time: 1500, count: 3, pattern: 1, ySpacing: 2000, xSpacing: 0 }]];